lib/rack/attack/throttle.rb

```diff
@@ -23,34 +23,48 @@ def cache
 
       def matched_by?(request)
         discriminator = block.call(request)
+
         return false unless discriminator
 
-        current_period = period.respond_to?(:call) ? period.call(request) : period
-        current_limit  = limit.respond_to?(:call) ? limit.call(request) : limit
-        key            = "#{name}:#{discriminator}"
-        count          = cache.count(key, current_period)
-        epoch_time     = cache.last_epoch_time
+        current_period  = current_period(request)
+        current_limit   = current_limit(request)
+        count           = cache.count("#{name}:#{discriminator}", current_period)
 
         data = {
           discriminator: discriminator,
           count: count,
           period: current_period,
           limit: current_limit,
-          epoch_time: epoch_time
+          epoch_time: cache.last_epoch_time
         }
 
-        (request.env['rack.attack.throttle_data'] ||= {})[name] = data
-
         (count > current_limit).tap do |throttled|
-          if throttled
-            request.env['rack.attack.matched']             = name
-            request.env['rack.attack.match_discriminator'] = discriminator
-            request.env['rack.attack.match_type']          = type
-            request.env['rack.attack.match_data']          = data
-            Rack::Attack.instrument(request)
-          end
+          annotate_request_with_throttle_data(request, data)
+          annotate_request_with_matched_data(request, data) if throttled
         end
       end
+
+      private
+
+      def current_period(request)
+        period.respond_to?(:call) ? period.call(request) : period
+      end
+
+      def current_limit(request)
+        limit.respond_to?(:call) ? limit.call(request) : limit
+      end
+
+      def annotate_request_with_throttle_data(request, data)
+        (request.env['rack.attack.throttle_data'] ||= {})[name] = data
+      end
+
+      def annotate_request_with_matched_data(request, data)
+        request.env['rack.attack.matched']             = name
+        request.env['rack.attack.match_discriminator'] = data[:discriminator]
+        request.env['rack.attack.match_type']          = type
+        request.env['rack.attack.match_data']          = data
+        Rack::Attack.instrument(request)
```

Comment:
I feel it might be a bit misleading or unexpected that a method named "annotate request with something" ends up sending a notification with the `instrument` method here.

What about leaving it outside?

Reply:
You're right, changed!